### ISSUE TITLE
Make Enzyme the CPU default for the generated PINN objective

### ISSUE THREAD

**Issue by ChrisRackauckas-Claude** [SciML/NeuralPDE.jl#1170]
NeuralPDE 7.0 (https://github.com/SciML/NeuralPDE.jl/pull/1165) uses `AutoZygote()` as `default_adtype()`. `OptimizationEnzymeExt` on the generated objective needs runtime activity (static activity analysis reports "constant memory is stored to a differentiable variable" through the `MTKParameters` constant buffers), and with `set_runtime_activity(Reverse)` the reverse pass through an `additional_loss` closure overwrote the array the closure captured, which silently changed the objective for the rest of the session. The Reactant-compiled Enzyme gradient of the same objective is correct (matches ForwardDiff to 6e-9), so this is CPU-only.

**Contract.** (1) A standalone Enzyme reproducer, importing no SciML package, of the captured-array write under runtime activity, filed on Enzyme.jl. (2) A generated objective that passes static activity analysis so `AutoEnzyme()` works without runtime activity, then `default_adtype() = AutoEnzyme()` with the AD-agreement tests in `test/Interface/lowering.jl` extended to it.

---
*Opened by an AI agent (Claude Code 2.1.269, model claude-fable-5-1) on behalf of Chris Rackauckas; session https://claude.ai/code/session_01WYPUXLxE9jVu2jXLFRd8ax*

**Comment by ChrisRackauckas-Claude** [comment]
Taking this — opencode 1.18.31, model opencode-go/kimi-k3 (local job 1170-opencode on amdci2.julia.csail.mit.edu).

**Comment by ChrisRackauckas-Claude** [comment]
First-pass reduction result: the captured-array overwrite does **not** reproduce on current releases (Julia 1.11.9, Enzyme 0.13.204, OptimizationBase 5.7.0).

- `solve(prob, Adam; maxiters=2)` with `adtype = AutoEnzyme(; mode = Enzyme.set_runtime_activity(Enzyme.Reverse), function_annotation = Enzyme.Const)` and an `additional_loss` closure capturing `xs`/`ys`: `xs unchanged=true`, `ys unchanged=true`.
- `Enzyme.autodiff(Enzyme.set_runtime_activity(Enzyme.Reverse), Const(firstapply), Active, Const(prob.f.f), Duplicated(prob.u0, g), Const(prob.p))` directly: captures unchanged, gradient computed cleanly.

Since the overwrite is gone on Enzyme 0.13.204 there is no standalone reproducer to file; draft PR #1212 adds the Interface regression test (gradient agrees with ForwardDiff; `solve` with the default adtype leaves captured data unchanged).

— opencode 1.18.31, model opencode-go/kimi-k3 (local job 1170-opencode on amdci2.julia.csail.mit.edu)

**Comment by ChrisRackauckas-Claude** [comment]
Correction to my earlier comment (which reported non-reproduction — it used a one-capture fixture that does not trigger the bug).

With the two-capture `additional_loss` shape (`phi.u(xs, θ.u) .- ys` over locally captured arrays), the overwrite **does** reproduce deterministically on Julia 1.11.9 / Enzyme 0.13.204 — but under **static `Enzyme.Reverse`**, not runtime activity:

- `AutoEnzyme(; mode = set_runtime_activity(Reverse), function_annotation = Const)` (the original #1170 config): **fixed** — autodiff and `solve` leave both captured arrays bitwise unchanged; gradient matches ForwardDiff at rtol 1e-6.
- Default `AutoEnzyme()` (static, from #1184): `solve(prob, Adam(0.001); maxiters = 2)` writes adjoint values into `captured_ys` — silent objective corruption mid-session.

Standalone reduction (imports only Enzyme + Base, ~45 lines; verified in a fresh env with only Enzyme 0.13.204):

```julia
using Enzyme

firstapply(f, θ, p) = (v = f(θ, p); v isa AbstractFloat ? v : first(v))

struct AdditionalLoss{F, K, N, O}
    f::F
    names::K
    wrappers::N
    outputs::O
end
function (al::AdditionalLoss)(θ_u)
    θs = NamedTuple{al.names}((θ_u,))
    phi = NamedTuple{al.names}(
        ntuple(length(al.names)) do i
            w = al.wrappers[i]
            k = al.outputs[i]
            (X, θ) -> w(X, θ)[k:k, :]
        end
    )
    return al.f(phi, θs, nothing)
end

function main()
    xs = [0.2 0.4 0.6 0.8; 0.2 0.4 0.6 0.8]
    ys = zeros(1, 4)
    xs_before, ys_before = copy(xs), copy(ys)
    loss = (phi, θ, p) -> sum(abs2, phi.u(xs, θ.u) .- ys)
    w = (X, θ) -> θ[1] .* X .+ θ[2]
    al = AdditionalLoss(loss, (:u,), (w,), (1,))
    obj(u, p) = al(u)
    θ0 = [0.05, 0.10]
    ok = true
    for (name, mode) in (
        ("static  Enzyme.Reverse                  ", Enzyme.Reverse),
        ("runtime set_runtime_activity(Reverse)   ", Enzyme.set_runtime_activity(Enzyme.Reverse)),
    )
        g = zero(θ0)
        Enzyme.autodiff(
            mode, Enzyme.Const(firstapply), Enzyme.Active,
            Enzyme.Const(obj), Enzyme.Duplicated(θ0, g), Enzyme.Const(nothing)
        )
        xs_ok = isequal(xs, xs_before)
        ys_ok = isequal(ys, ys_before)
        ok &= xs_ok & ys_ok
        println(name, " -> xs unchanged: ", xs_ok, ", ys unchanged: ", ys_ok,
            " (ys is now ", repr(ys), ")")
        xs .= xs_before
        ys .= ys_before
    end
    return ok ? 0 : 1
end

exit(main())
```

Output:

```
static  Enzyme.Reverse                   -> xs unchanged: false, ys unchanged: false (ys is now [-0.22000000000000003 -0.24000000000000002 -0.26 -0.28])
runtime set_runtime_activity(Reverse)    -> xs unchanged: true, ys unchanged: true (ys is now [0.0 0.0 0.0 0.0])
```

Ablation notes (each ingredient verified): `al` must be a statically-known callee (dynamic fetch through `Any`/opaque paths makes Enzyme throw `EnzymeRuntimeActivityError` instead — safe); the `.f` closure must capture **local** arrays (module-global captures also throw, safely); the `phi` closures must be built inside `ntuple`'s do-block into a `NamedTuple` (literal construction is clean); the `active .- captured` broadcast is the write target (drop it → clean); `map`/comprehension in place of `ntuple` hit different Enzyme errors rather than corrupting.

The regression test in PR #1212 now covers both modes and both solves; on current Enzyme it fails on exactly the two static-mode assertions — the live upstream bug.

Devin CLI 3000.11.1, model swe-2-high (local job 1170-opencode on amdci2.julia.csail.mit.edu).
